File: cse-analyzer/backend/app/routers/screener.py
from typing import Optional
from fastapi import APIRouter, Depends, Query
from app.middleware.auth_middleware import get_current_user
from app.services import screener_service

router = APIRouter(prefix="/screener", tags=["screener"])


@router.get("")
async def screen_stocks(
    sector:         Optional[str]   = Query(None),
    min_price:      Optional[float] = Query(None),
    max_price:      Optional[float] = Query(None),
    min_change_pct: Optional[float] = Query(None),
    max_change_pct: Optional[float] = Query(None),
    min_volume:     Optional[float] = Query(None),
    min_market_cap: Optional[float] = Query(None),
    max_market_cap: Optional[float] = Query(None),
    near_52w_high:  Optional[bool]  = Query(None),
    near_52w_low:   Optional[bool]  = Query(None),
    sort_by:        str             = Query("symbol"),
    sort_dir:       int             = Query(1),
    limit:          int             = Query(200, le=500),
    current_user: dict = Depends(get_current_user),
):
    return await screener_service.run_screener(
        sector=sector,
        min_price=min_price,
        max_price=max_price,
        min_change_pct=min_change_pct,
        max_change_pct=max_change_pct,
        min_volume=min_volume,
        min_market_cap=min_market_cap,
        max_market_cap=max_market_cap,
        near_52w_high=near_52w_high,
        near_52w_low=near_52w_low,
        sort_by=sort_by,
        sort_dir=sort_dir,
        limit=limit,
    )


@router.get("/sectors")
async def list_sectors(current_user: dict = Depends(get_current_user)):
    return await screener_service.get_sectors()

File: cse-analyzer/frontend/src/api/screener.ts
import client from "./client";

export interface ScreenerFilters {
  sector?: string;
  min_price?: number;
  max_price?: number;
  min_change_pct?: number;
  max_change_pct?: number;
  min_volume?: number;
  min_market_cap?: number;
  max_market_cap?: number;
  near_52w_high?: boolean;
  near_52w_low?: boolean;
  sort_by?: string;
  sort_dir?: number;
  limit?: number;
}

export interface ScreenerRow {
  symbol: string;
  name?: string;
  sector?: string;
  close?: number;
  open?: number;
  high?: number;
  low?: number;
  pct_change?: number;
  volume?: number;
  turnover?: number;
  market_cap?: number;
  high_52w?: number;
  low_52w?: number;
  date?: string;
}

export const runScreener = (filters: ScreenerFilters = {}) => {
  const params = Object.fromEntries(
    Object.entries(filters).filter(([, v]) => v !== undefined && v !== "" && v !== null)
  );
  return client.get<ScreenerRow[]>("/screener", { params }).then(r => r.data);
};

export const getScreenerSectors = () =>
  client.get<string[]>("/screener/sectors").then(r => r.data);
